sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -264,6 +264,10 @@ object NullPropagation extends Rule[LogicalPlan] {
       case e @ Substring(_, Literal(null, _), _) => Literal.create(null, e.dataType)
       case e @ Substring(_, _, Literal(null, _)) => Literal.create(null, e.dataType)
 
+      // MaxOf and MinOf can't do null propagation
```

Comment:
`MaxOf` and `MinOf` are not `BinaryArithmetic` before, which leads to a lot of duplicated codes. So I made them extend `BinaryArithmetic` and add them as exceptional cases here.